src/coreclr/jit/emit.cpp

```diff
@@ -4102,6 +4122,315 @@ void emitter::emitDispCommentForHandle(size_t handle, GenTreeFlags flag)
+                    {
+                        CLANG_FORMAT_COMMENT_ANCHOR
+                        // the last instruction in the group is the jmp we're looking for
+                        // and it jumps to the next instruction group so we don't really need it
+
+                        if (!previousInstructionWasCall)
+                        {
+#ifdef DEBUG
+                            if (EMITVERBOSE)
+                            {
+                                printf("Removing unconditional jump [%08X/%03u] from the last instruction in "
+                                       "IG%02u to the next IG IG%02u label %s\n",
+                                       dspPtr(jmp), jmp->idDebugOnlyInfo()->idNum, group->igNum, targetGroup->igNum,
+                                       emitLabelString(targetGroup));
+                            }
+#endif
+                            UNATIVE_OFFSET sizeRemoved = instructionDescriptor->idCodeSize();
+
+                            // set state needed at the end of the loop in ADJUST_GROUP
+                            adjIG      = sizeRemoved;
+                            lstIG      = group;
+                            jmpRemoved = true;
+
+                            // unlink the jump
+                            if (previousJmp != nullptr)
+                            {
+                                previousJmp->idjNext = jmp->idjNext;
+                                if (jmp == emitJumpLast)
+                                {
+                                    emitJumpLast = previousJmp;
+                                }
+                            }
+                            else
+                            {
+                                assert(jmp == emitJumpList);
+                                emitJumpList = jmp->idjNext;
+                            }
+                            jmp->idjNext = nullptr;
+
+#if DEBUG
+                            // clear the instruction data
+                            memset((BYTE*)instructionDescriptor, 0, instructionDescriptor->idCodeSize());
+#endif
+
+                            // remove the instruction from the group
+                            group->igInsCnt -= 1;
+                            group->igSize -= (unsigned short)sizeRemoved;
+                            emitTotalCodeSize -= sizeRemoved;
+
+                            // flag the group as resized and recounted
+                            group->igFlags |= (IGF_UPD_ISZ | IGF_UPD_ICOUNT);
+
+                            // cleanup
+                            instructionDescriptor = nullptr;
+                            jmp                   = nullptr;
+                            removedCount += 1;
+
+                            goto ADJUST_GROUP;
+                        }
+#ifdef DEBUG
+                        else
+                        {
+                            // due to x64 ABI semantics we can't remove the jmp if the IG ends
```

Comment:
This description is not quite precise. See https://github.com/dotnet/runtime/blob/8b8229b300536b8594227de282882a9f6c1dd18a/src/coreclr/jit/codegenlinear.cpp#L619-L627. It's also unfortunate to prevent this optimization (for all platforms) due to this not-quite-precise definition. It's also unfortunate to add this complex logic in yet another location.

Reply:
I would have preferred to find a way to mark the node in `genCodeForBBlist` but because it is strictly forward only there isn't a way to do it. We also don't have EH region code at that point we we can't mark the next instruction because it may not exist yet. 

I don't like how much extra logic it adds to try and backtrack to a previous IG but that case only occurred in 3 dedicated jit tests, everything else was caught by the logic added into the already required instruction walk through the ig.

Reply:
Can't use use `emitIsLastInsCall()` information at the time of setting `_idIsJmpAlways ` and then skip the entire logic in this method to detect if the previous instruction was a call?

So `emitIns_J` would be:

```c++
if (!emitIsLastInsCall()) {
  id->idSetJmpAlwaysFlag(isJmpAlways);
} else {
 id->idSetJmpAlwaysFlag(false);
}
```

Reply:
I had thought `emitIsLastInsCall()` was per ig but now i've checked it's per function, so yes I could use it. 

Using it would prevent me being able to see `call, jmp(always)` pairs though and I had hoped to be able to be able to do further work in another PR to transform those from `call, jmp` to `call, nop`. I consider transforming an instruction more error prone than removing one entirely but that seems to be different to the opinion here.

Reply:
This is implemented.